vault/logical_system_test.go

```diff
@@ -2730,3 +2733,662 @@ func TestSystemBackend_PathWildcardPreflight(t *testing.T) {
+			assertHasKey(t, actualResp.Data, logical.HTTPRawBody, "key %s not found in data", logical.HTTPRawBody)
+			assertIsString(t, actualResp.Data[logical.HTTPRawBody], "key %s should have a string value", logical.HTTPRawBody)
+			rawBody := actualResp.Data[logical.HTTPRawBody].(string)
+
+			// Delete the body so the rest of the response can be compared
+			delete(actualResp.Data, logical.HTTPRawBody)
+			assert(t, reflect.DeepEqual(actualResp, expectedResp), "Actual response: %#v\nExpected response: %#v", actualResp, expectedResp)
+
+			actualBody := &logical.HTTPResponse{}
+			err = json.Unmarshal([]byte(rawBody), actualBody)
+			assert(t, err == nil, "unable to unmarshal response body: %s", err)
+
+			assert(t, actualBody.Data != nil, "expected data field in body, got nil")
+			assertHasKey(t, actualBody.Data, "password", "expected data in body to have 'password' field")
+			assertIsString(t, actualBody.Data["password"], "password field should be a string")
+			password := actualBody.Data["password"].(string)
+
+			// delete the password field so the rest of the body can be compared
+			delete(actualBody.Data, "password")
+			assert(t, reflect.DeepEqual(actualBody, expectedBody), "Actual body: %#v\nExpected body: %#v", actualBody, expectedBody)
+
+			// Check to make sure the password is correctly formatted
+			passwordLength := len([]rune(password))
+			if passwordLength != expectedPassLen {
+				t.Fatalf("password is %d characters but should be %d", passwordLength, expectedPassLen)
+			}
+
+			for _, rule := range rules {
+				if !rule.Pass([]rune(password)) {
+					t.Fatalf("password %s does not have the correct characters", password)
+				}
+			}
+		}
+	})
+}
+
+func toHTTPResponse(t *testing.T, code int, resp *logical.Response) *logical.Response {
+	t.Helper()
+
+	httpResp, err := logical.RespondWithStatusCode(resp, nil, code)
+	if err != nil {
+		t.Fatalf("failed to convert logical response to HTTP response: %s", err)
+	}
+	return httpResp
+}
+
+func assert(t *testing.T, pass bool, f string, vals ...interface{}) {
+	t.Helper()
+	if !pass {
+		t.Fatalf(f, vals...)
+	}
+}
+
+func assertHasKey(t *testing.T, m map[string]interface{}, key string, f string, vals ...interface{}) {
+	t.Helper()
+	_, exists := m[key]
+	if !exists {
+		t.Fatalf(f, vals...)
+	}
+}
+
+func assertIsString(t *testing.T, val interface{}, f string, vals ...interface{}) {
+	t.Helper()
+	_, ok := val.(string)
+	if !ok {
+		t.Fatalf(f, vals...)
+	}
+}
+
+func passwordPoliciesFieldData(raw map[string]interface{}) *framework.FieldData {
+	return &framework.FieldData{
+		Raw: raw,
+		Schema: map[string]*framework.FieldSchema{
+			"name": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "The name of the password policy.",
+			},
+			"policy": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "The password policy",
+			},
+		},
+	}
+}
+
+type fakeStorage struct {
```

Comment:
We already have some helpers for things like this, consider extending those interested of introducing a new mock here. Checkout the sdk/physical directory

Reply:
I looked into the existing fakes, but decided against using them here. The use case is different enough that I decided to switch this over to a mock from gomock.

Reply:
Switched this over to using our existing fakes with some new test helpers.